W5500.device.nut

```diff
@@ -0,0 +1,2538 @@
+        _driver.openConnection(ip, port, mode, cb);
+    }
+
+    // ***************************************************************************
+    // listen
+    // Returns: this
+    // Parameters:
+    //      sourcePort - the port to listen on
+    //      cb - function to be called when connection successfully established 
+    // ****************************************************************************
+    function listen(sourcePort, cb) {
+        if (!_isReady) throw "Wiznet driver not ready";
+
+        // Create a socket, a connection object & a handler
+        _driver.listen(sourcePort, cb);
+    }
+
+    // ***************************************************************************
+    // getNumSockets - 
+    // Returns: returns the total number of sockets available
+    // Parameters:
+    //      none
+    // **************************************************************************
+    function getNumSockets() {
+        return _driver._noOfSockets;
+    }
+
+
+    // ***************************************************************************
+    // getNumSocketsFree - 
+    // Returns: returns the number of unused sockets
+    // Parameters:
+    //      none
+    // **************************************************************************
+    function getNumSocketsFree() {
+        return _driver._noOfSockets - _driver._connections.len();
+    }
+
+
+}
+
+
+// ==============================================================================
+// CLASS: W5500.Driver
+// ==============================================================================
+
+class W5500.Driver {
+
+    // Chip can support 8 sockets
+    static TOTAL_SUPPORTED_SOCKETS = 8;
+    static MAX_TX_MEM_BUFFER = 16;
+    static MAX_RX_MEM_BUFFER = 16;
+
+    // CLASS VARIABLES
+    _interruptPin = null;
+    _spi = null;
+    _cs = null;
+    _resetPin = null;
+
+    _connections = null; // open connections
+    _availableSockets = null; // array of sockets available
+    _maxNoOfSockets = 0; // max number of sockets wiznet may use
+    _noOfSockets = 0; // number of sockets
+    _socketMemory = null; // amount of memeory each socket has
+
+    // ***************************************************************************
+    // Constructor
+    // Returns: null
+    // Parameters:
+    //      spi - configured spi, W5500 supports spi mode 0 or 3
+    //      cs(optional) - configured chip select pin
+    //      reset(optional) - configured reset pin
+    // ***************************************************************************
+    constructor(interruptPin, spi, cs = null, resetPin = null) {
+
+        _interruptPin = interruptPin;
+        _spi = spi;
+        _cs = cs;
+        _resetPin = resetPin;
+        _connections = {};
+        _availableSockets = [];
+
+        // Check the pins
+        if (resetPin) _resetPin.configure(DIGITAL_OUT, 1);
+
+        local imp005 = ("spi0" in hardware);
```

Comment:
Please add a To Do comment here -> determine device type with imp.info() when realease 36 is default impOS version